### ISSUE TITLE
rdocumentation on 2.2?

### ISSUE THREAD

**Issue by zmjones** [berndbischl/mlr#330]
As I was working on the tutorial I noticed [rdocumentation](http://www.rdocumentation.org/packages/mlr) shows the docs from version 2.2. None of the new functions seem to have working links.

**Comment by berndbischl** [comment]
I am not sure how often they update. We used a different plattform before, they also did not update enough.

**Comment by zmjones** [comment]
So this brings up another tutorial related question I have. Why do we have a separate tutorial and not a series of vignettes? The bad thing about this is that the tutorial people see does not match the current cran version, which I would guess is confusing.

**Comment by schiffner** [comment]
About rdocumentation: They say that they automatically update documentation for CRAN packages. But that does not seem to be working. I tried to contact them twice already, but got no response. I was going to look for another solution this week.

**Comment by berndbischl** [comment]
a) It is horrible to interlink between parts in vignettes

b) PDFs are much worse than HTML (IMHO)

Why does the tutorial not match CRAN? it should

**Comment by schiffner** [comment]
Normally, the tutorial is ahead of CRAN, because otherwise we would have to wait with updates until after the CRAN release.

**Comment by zmjones** [comment]
Well can't we do html vignettes now? e.g. like [dplyr](http://cran.r-project.org/web/packages/dplyr/index.html)

**Comment by berndbischl** [comment]
We can but have you checked how to link to other parts?

**Comment by berndbischl** [comment]
Also, I dont think we should discuss changing the complete structure now

**Comment by zmjones** [comment]
No I haven't, fair enough.

**Comment by berndbischl** [comment]
I emailed rdoc again. We cannot do much here except wait? Close?

**Comment by jakob-r** [comment]
@schiffner We could work with branches on gh-pages as well.

**Comment by berndbischl** [comment]
Yeah, maybe we could do this. We could have a sperate gh-pages-devel branch. We merge on only on CRAN push.

@schiffner does the most work here. We should follow her gut feeling and decision.

**Comment by schiffner** [comment]
I guess we could do this.
New mlr features go into gh-pages-devel, all other improvements (at least major ones like new sections about older features or error corrections) should preferably go into both gh-pages and gh-pages-devel. I am currently finding out on how to do the latter comfortably. If anyone has experience with this, let me know.

**Comment by berndbischl** [comment]
Create that branch ghpages-devel and copy the current state into it.
If updates are done to ghpages, merge them also into the devel branch as soon as you can.

Is that about correct?
@larskotthoff  @mllg ?

**Comment by berndbischl** [comment]
And when we release to cran you merge the devel branch into the normal one.

**Comment by jakob-r** [comment]
I think it is save to have the two branches like you mentioned. If you want to fix something on gh-pages it will be something like
```
git checkout gh-pages
git pull
# change stuff and add it
git commit -m "fix on master gh-pages"
git push
git checkout gh-pages-devel
# now get the changes from gh-pages
git merge origin/gh-pages
# hope for no conflicts
git push
```
If we have a new version on cran we just merge the other way round. Should work fine as we don't have to many conflicting edits in the tutorial.

PS: Big thanks for maintaining the tutorial!

**Comment by larskotthoff** [comment]
I'd prefer a way to make the -devel branch visible as well, maybe with different top-level directories (i.e. tutorial/current/html, tutorial/2.2/html, etc). We wouldn't get any particularly useful merging between the directories, but do we even need that? I.e. when a new mlr release is made, -current becomes -2.5 or whatever and is never changed. All changes happen in -current.

**Comment by berndbischl** [comment]
Do you now want to maintain the old dirs and apply bugfixes to them?

**Comment by berndbischl** [comment]
But in general I like the idea, I guess. Because people could access "older states" if they have an older version. But we dont maintain those. 

So what we write always goes to "current". That gets copied / freezed when we release to CRAN to a numbered dir. 

In that case we dont even need branches for the RMD files. We just link to "current" and "2.4" (the latest CRAN release) on the webpage and explain what it is. No merging, etc. 
We just copy the hmtl at a certain point in time.

Is this best?

**Comment by larskotthoff** [comment]
I wouldn't. We don't do that for old releases either, do we?

**Comment by berndbischl** [comment]
Read my proposal above.

**Comment by larskotthoff** [comment]
And yes, that's exactly what I have in mind.

**Comment by berndbischl** [comment]
@schiffner 
Your 2 cents?

**Comment by schiffner** [comment]
I agree.
The only thing I was thinking about earlier this day was if one could efficiently maintain both the CRAN release (2.4) and the development version because many of (at least my fixes) are not about new mlr features. Something like always pushing to gh-pages-devel and cherry-picking commits for gh-pages with no merging needed...
But I guess this can get complicated and just freezing the html is a better solution, particularly if the html files for CRAN release and devel are both visible.

**Comment by berndbischl** [comment]
Yeah. Most people will refer to "current". But if they get confused, because some minor thing was changed in the API, we can simply point them to 2.4. 
And the tutorial will continuously get better. 
And nobody will make mistakes due to a complicated setup.

@larskotthoff 
Could you be so nice to look at the generating scripts so what you proposed can be done?

**Comment by larskotthoff** [comment]
I was afraid you were going to say something like that :) I'll have a look.

**Comment by larskotthoff** [comment]
Ok, it looks like this should be reasonably easy, although we'll have to manually copy the tutorial for each release.

I can't compile the tutorial at the moment though, getting

    Knitting file 'benchmark_experiments.Rmd' ...
    Note: the specification for class "numeric or NULL" in package 'clValid' seems equivalent to one from package 'SparseM': not turning on duplicate class definitions for this class.
    Quitting from lines 128-143 ()
    Error in makeLogicalLearnerParam(id = "CV", default = FALSE, tunable = FALSE) :
      unused argument (tunable = FALSE)
    Calls: lapply ... checkClass -> makeParamSet -> makeLogicalLearnerParam
    Execution halted

with HEAD and

    Knitting file 'benchmark_experiments.Rmd' ...
    Note: the specification for class "numeric or NULL" in package 'clValid' seems equivalent to one from package 'SparseM': not turning on duplicate class definitions for this class.
    Quitting from lines 128-143 ()
    Error in FUN(X[[i]], ...) :
      List measures has element of wrong type function at position 2. Should be: Measure
    Calls: lapply ... benchmark -> checkListElementClass -> lapply -> FUN
    Execution halted

Am I missing something to make this work?

**Comment by schiffner** [comment]
This problem is new to me. I will have a look.

Wouldn't is be better, in order to freeze the tutorial for mlr 2.4, to build it with the CRAN release version (or am I missing something)?

**Comment by berndbischl** [comment]
You need to update ParamHelpers for now from GH

**Comment by berndbischl** [comment]
and freeze it for the 2.4, yes!

**Comment by larskotthoff** [comment]
Yes, that was the plan. Fixed for HEAD now, but I'm getting the same error I'm getting for 2.4 now.

**Comment by berndbischl** [comment]
Can you copypaste the lines here that get execed?

**Comment by schiffner** [comment]
I just checked. I can build the benchmark section of the tutorial without problems with mlr 2.4 and ParamHelpers HEAD.

Have you deleted the cache? No offense intended, just asking because I run into this trap fairly often.

**Comment by berndbischl** [comment]
auc is probably "shadowed" again. that is why checkmate is nice. It clearly says that.

We will move the tutorial to a new repo. I have create mlr-org as an orga for that. 

There we will build a travis job to always check it.

@larskotthoff : We need to make sure that a push to the mlr repo induces a build for the tutorial....
Well we dont need to but it would be helpful

**Comment by larskotthoff** [comment]
@schiffner Thanks, that was it!

**Comment by schiffner** [comment]
@larskotthoff Great! (Sorry for the misunderstanding earlier, just realized after reading a second time that you had already planned to build the tutorial with both HEAD and 2.4...).
Are you working on the tutorial stuff right now? I'm doing some proofreading at the moment (nothing mlr 2.5 related). Just let me know if I should hold off on commits.

**Comment by larskotthoff** [comment]
I'm just moving it to the new location, should be done in a few minutes.

**Comment by larskotthoff** [comment]
Ok, the tutorial now has its own repository at https://github.com/mlr-org/mlr-tutorial/tree/gh-pages -- please use that to commit any changes. The new website is https://mlr-org.github.io/mlr-tutorial/tutorial/current/html/ (I haven't done the "archive" one for 2.4 yet). I'll have a look at the travis job now.

**Comment by schiffner** [comment]
Thanks!

**Comment by larskotthoff** [comment]
Ok, the version for 2.4 is now available at https://mlr-org.github.io/mlr-tutorial/tutorial/2.4/html/

**Comment by schiffner** [comment]
Thanks again.
I had a look with several different browsers. Do you also have the problem that text in plots like axis labels or panel labels appears a bit blurred? Moreover, at least for me, the formulas in https://mlr-org.github.io/mlr-tutorial/tutorial/current/html/cost_sensitive_classif/index.html are not displayed correctly.

**Comment by larskotthoff** [comment]
The text looks ok to me, but the images were indeed broken -- the links are actually hardcoded. This should be fixed now, thanks for checking. The links probably shouldn't be hardcoded though, I've opened an [issue](https://github.com/mlr-org/mlr-tutorial/issues/1) in the new repo.

**Comment by jakob-r** [comment]
We use the vector based image format svg (which's use I strongly favour). Depending on your build system the text is generated using some ugly pixel font like you see now which is kind of a fallback option.

**Comment by zmjones** [comment]
Is there a reason we are avoiding mathjax?

**Comment by larskotthoff** [comment]
@jakob-r So what do I need to install to make this work properly?

**Comment by schiffner** [comment]
I'm currently reading about it.
What I know is that you need to install an extension for Python markdown. There seems to be no officially supported extension (http://pythonhosted.org/Markdown/extensions/index.html), but some third-party extensions for math (https://github.com/waylan/Python-Markdown/wiki/Third-Party-Extensions). One of them is for mathjax support (https://github.com/mitya57/python-markdown-math). The README says after installation the extension name needs to be added to mkdocs.yml (http://www.mkdocs.org/user-guide/configuration/#formatting-options), and some lines for including mathjax and configuration need to be added.
I can continue and try to make it work, unless you, Lars or Jakob, want to do this.

**Comment by larskotthoff** [comment]
@schiffner It would be great if you could look into this -- I'm still trying to get the Travis build to work. Once that is done, I'll see if we can have Travis update the tutorial so that everything is consistent and doesn't change depending on who last updated it.

**Comment by larskotthoff** [comment]
Ok, the travis build is somewhat working now (see https://travis-ci.org/mlr-org/mlr-tutorial/builds/67405428). It fails saying that `clusterSim` isn't installed (which it is). I get the same error message locally as well, which disappears when I just run `./build` again. Does anybody know what's going on here?

**Comment by zmjones** [comment]
i see that error frequently as well

**Comment by larskotthoff** [comment]
Upon further investigation it seems to be related to the maximum number of shared libraries that can be loaded into R at any one time. This limit is fixed at compile time. I guess we could unload all loaded libraries after each file, but that would probably be a huge pain.

**Comment by berndbischl** [comment]
> Upon further investigation it seems to be related to the maximum number of shared libraries

This sucks so fucking much

**Comment by berndbischl** [comment]
And unloading is not guaranteed to work

**Comment by berndbischl** [comment]
But: Why do you load so much when building the tutorial? Just for the learner table?

**Comment by larskotthoff** [comment]
I've changed the build to retry if failed now. The first time it fails is when compiling the predict part -- the tables of learners, measures etc are before that, so it's possible that it's mostly caused by them.

**Comment by schiffner** [comment]
@jakob-r : Thanks for clearing things up about the graphics.
@larskotthoff : Thanks for migrating the tutorial and making travis work.

@zmjones : Thanks for reminding me about mathjax. I had a short look at it sometime last year, but there were more pressing matters, therefore I resorted to the LaTeX->image solution.

The mathjax stuff seems to be working now. You just need to install a math extension for Python markdown. I put all infos in the README. Let me know if it works and if you like it.

**Comment by zmjones** [comment]
Yep! I use it for my website so it must be pretty easy ha.

**Comment by jakob-r** [comment]
@larskotthoff I had this issue once really long time ago. For now I also don't find anything relating to that issue. Whoever runs into that issue maybe has to check his logs.

**Comment by larskotthoff** [comment]
The tutorial is now (after many iterations) automatically updated and pushed by Travis. This means that you don't need to generate everything yourself anymore when making changes. Please check whether anything looks weird.

**Comment by jakob-r** [comment]
Wow. Excellent (miss)usage of travis :grinning: 
current looks perfectly fine. 2.4 still has the font issue

**Comment by berndbischl** [comment]
I had 0 time to look at this until now. But thx already :)

**Comment by larskotthoff** [comment]
@jakob-r Even more so as first I had produced an infinite loop of Travis triggering itself through the commit ;) If you could regenerate 2.4 and commit that would be great -- I don't know what to install/change to fix this on my machine.

**Comment by schiffner** [comment]
@larskotthoff Thanks for making travis work.

@jakob-r @larskotthoff I can regenerate 2.4 and commit. I don't have any font issues.

**Comment by schiffner** [comment]
I pushed a new version of 2.4. Please check if everything looks ok.

**Comment by mllg** [comment]
I'd recommend to use "stable" and "devel" as directory names (for bookmarks etc.). Fonts seem to be fixed.

**Comment by schiffner** [comment]
Agreed.

**Comment by schiffner** [comment]
Would "release" and "devel" also be ok?

**Comment by berndbischl** [comment]
Sounds also good to me

**Comment by mllg** [comment]
Sure. Everything but the version number is fine. :+1:

**Comment by berndbischl** [comment]
And what do we do with older versions? I though the idea was to store these as well?
(I can live with the fact that we dont do this. But it sounded nice)

**Comment by larskotthoff** [comment]
So have the symbolic names "stable" for the current CRAN version (symlinked to the corresponding version number) and "devel" in addition to the directories for specific CRAN version numbers?

**Comment by berndbischl** [comment]
Yes I would like that, it it is easy to do

**Comment by schiffner** [comment]
+1

@berndbischl : Never meant to not store the older versions. Just wanted to avoid maintaining links in the mlr README / Vignette that contain version numbers. :)

**Comment by schiffner** [comment]
Arrgh, I just realized that I messed up the image links in 2.4. Will fix that tomorrow...

**Comment by larskotthoff** [comment]
Ok, I've changed "current" to "devel" and add a "release" symlink.

**Comment by berndbischl** [comment]
The current "release" link is a 404?

**Comment by berndbischl** [comment]
and why are we using this old and closed thread

**Comment by larskotthoff** [comment]
Ok, apparently you have to disable jekyll to support symlinks. This should work now.

Also, feel free to open a new issue in the new repo :)

**Comment by schiffner** [comment]
Thanks @larskotthoff .

Just to wrap this up. The image links in 2.4 (release) should work now.
The images in the zips (release and devel) are broken (and I doubt that this ever worked in the old repo unless I'm missing something big time...), because we don't put them in there.
I will have a closer look and open an issue in the new repo if necessary.
